ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -419,115 +402,28 @@ public boolean hasSucceeded() {
-      // 2. (allReplicaCount - totalNotFoundCount >= replicaSuccessTarget)
-      //    Minus all the NOT_FOUND replicas, among all the replicas belongs to the partition(no matter online or offline),
-      //    we should still have enough replicas to satisfy the replicaSuccessTarget.
-      //    That means, if we retry, potentially we have the chance to be successful and get the blob back from other replica.
-      // 3. (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1)
-      //    The retry can be successful only if there is some offline replica coming back when retry happens.
-      //    That's why for TtlUpdate and Delete, we have the third condition as (totalOfflineReplicaCount > 0)
-      //    a). For GetBlob, the difference is that we already include OFFLINE replicas to the eligible replica pool
-      //        while TTLUpdate/Delete don't. Refer to routerOperationTrackerIncludeDownReplicas related code.
-      //        So it has tried the OFFLINE replicas already.
-      //    b). But with the recent Helix implementation, when the server is crashed,
-      //        the helix external view sometimes doesn't include the replica at all, not even in the OFFLINE category.
-      //        So we check how many replicas from the originating DC we read already.
-      //        Since we put routerPutSuccessTarget copies in the originating DC,
-      //        we should at least read "originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1" replicas
-      //        to make sure we read one of the copies.
-      //        If the eligible originating replica count is less that that, we should retry.
-      if (hasFailedOnCrossColoNotFound()
-          && (allReplicaCount - totalNotFoundCount >= replicaSuccessTarget)
-          && (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1)) {
-        logger.info(
-            "Terminating {} on {} due to disk down count and total Not_Found count from eligible replicas and some "
-                + "other replicas being unavailable. CrossColoEnabled: {}, DiskDownCount: {}, TotalNotFoundCount: {}, "
-                + "TotalReplicaCount: {}, replicaSuccessTarget: {}, OfflineReplicaCount: {}, allReplicaCount: {}, "
-                + "originatingDcEligibleReplicaCount {} {} "
-                + "replicasByState = {}",
-            routerOperation, partitionId, crossColoEnabled, diskDownCount, totalNotFoundCount, totalReplicaCount,
-            replicaSuccessTarget, totalOfflineReplicaCount, allReplicaCount, originatingDcEligibleReplicaCount,
-            getBlobIdLog(), allDcReplicasByState);
-        routerMetrics.failedMaybeDueToTotalOfflineReplicasCount.inc();
-        return true;
-      }
-      return false;
+    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
```

Comment:
Threshold being valid doesn't mean the result should be inconclusive. There could be a case where the operation is terminated before any of the request is sent out, like quota rejection. In this case, we have no NOT_FOUND error returned from server and the threshold could be -1 as well. In this way, we should return false.

I guess it's better in this way
```
if (originatingDcNotFoundFailureThreshold != NOT_FOUND_THRESHOLD_INVALID_ALUE) {
    if (originatingDcNotFoundCount >= originatingDcNotFoundFailureThreshold) {
        return True;
    }
    return False;
} else {
    if (originatingDcNotFoundCount == availableReplicasInOrginatingDC) {
        return INCONCLUSIVE; // This means even if all the available replicas returns NOT_FOUND, we still can't say it's not FOUND.
    }
    return False; // As long as there is no replica not returning NOT_FOUND
}
```

Reply:
Justin, I think the difference between your solution and Arun's solution is the case when "originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_ALUE".

In Arun's implementation, the INCONCLUSIVE stands for there is either some offline or bootstrapping originating replicas. Although for the offline replica, we didn't send any request but we should consider there is some "virtual offline status" from these replicas not existing in the external view. 
Because of that, in the Operation code, it's a "setOperationException(AmbryUnavailable)" to compensate the missing offline we should have but we didn't receive. It's not operationException.set(AmbryUnavailable) which will overwrite the final status.
 If there is any higher precedence error, the high precedence error still take the precedence and won't change anything. So in your case, it'll still return quota_rejected.


We don't have maybeFailedDueToOfflineReplicas any more in the checkAndMaybeComplete. So if we change to below, in the case for PutBlob, that we have 1 originating replica returns success and 8 replicas are offline, not sure what the status we will return. 
    if (originatingDcNotFoundCount == availableReplicasInOrginatingDC) {
        return INCONCLUSIVE; // This means even if all the available replicas returns NOT_FOUND, we still can't say it's not FOUND.
    }

Reply:
I don't think this is true. Check out the DeleteOperation::checkAndMaybeComplete method. when it's quota rejected, and not enough leader and standby in originating dc, it will return 503. But we want it to return  429.

Reply:
Yeah, actually like Jing was mentioning, I actually set 503 `setOperationException` so that errors which have more precedence are not replaced. Please let me know if this is okay.

Reply:
For me, I kind of feel it's easier to understand the code logic when we have two separated function, hasFailedOnNotFound and hasOfflineReplicas.
FailedOnNotFoundResult.INCONCLUSIVE basically means we have offline replica(or bootstrapping). We can use standalone hasOfflineReplicas() function. 
When hasOfflineReplicas is true and  hasFailedOnNotFound is false, we just fake one unavailable exception. But we honor the exception precedence and honor all the response we received.
Just want to know your opinion of it. 


```
if (operationTracker.hasFailedOnNotFound()) {
    if (originatingDcNotFoundFailureThreshold == NOT_FOUND_THRESHOLD_INVALID_VALUE) {
        operationException.set(RouterErrorCode.BlobDoesNotExist));
} else if (operationTracker.hasOfflineReplica()) {
      // compensate the offline status we should have but we don't offline is not in the clustermap
       setOperationException(RouterErrorCode.AmbryUnavailable):
}
```